### PR TITLE
Prepend "SVPNCOOKIE=" to the given cookie if missing

### PR DESCRIPTION
**Pull request by gm-vm**
It is maybe not obvious that currently the cookie is only accepted if it has "SVPNCOOKIE=", so add it when it is missing.

### CODE REVIEW

**Comment by gm-vm** [comment]
Sorry for force-pushing this a couple of times, I couldn't make my mind on how to handle the errors of `get_cookie_with_prefix` and didn't want to add useless commits.

I think that printing an error message and continuing is the easiest way to handle this, the authentication will simply fail and the reason should be clear.

**Comment by gm-vm** [comment]
Exiting immediately is not that more complicated than ignoring (the very unlikely) error, so I just did that.

I will squash the changes if this looks OK.

**Comment by gm-vm** [comment]
By the way, the last fix probably needs to be replicated for all the `strdup(optarg)`, but I think that needs to be done in a separate PR.

**Comment by DimitriPapadopoulos** [comment]
> By the way, the last fix probably needs to be replicated for all the `strdup(optarg)`, but I think that needs to be done in a separate PR.

@gm-vm Is #1041 what you had in mind?

**Comment by gm-vm** [comment]
Pretty much, yeah. I also considered adding something like the following and use it in both `config.c` and `main.c`:

```C
void set_string_argument(char **destination, const char *argument)
{
	free(*destination);
	*destination = strdup(argument);
	if (*destination == NULL)
		log_error("Could not set argument: %s\n", argument);
}
```

File: src/main.c [hunk 1]
```diff
@@ -517,19 +517,37 @@ int main(int argc, char **argv)
...
-				cli_cfg.cookie = strdup(optarg);
+				char *cookie_with_prefix;
+
+				cookie_with_prefix = strdup_with_prefix(optarg, "SVPNCOOKIE=");
+				if (cookie_with_prefix == NULL) {
+					log_error("Could not prepend \"SVPNCOOKIE=\" to the cookie.\n");
+					goto exit;
+				}
+				free(cli_cfg.cookie);
+				cli_cfg.cookie = cookie_with_prefix;
```
Comment (DimitriPapadopoulos) [inline src/main.c:528]: I would rather keep this PR focused on prepending SVPNCOOKIE, without adding checks:
````c
				free(cli_cfg.cookie);
				cli_cfg.cookie = strdup_with_prefix(optarg, "SVPNCOOKIE=");
````
I'm not saying these checks are bad. Just keep them for a different PR.

File: src/main.c [hunk 2]
```diff
@@ -517,19 +517,37 @@ int main(int argc, char **argv)
...
+				cookie_with_prefix = strdup_with_prefix(cookie, "SVPNCOOKIE=");
+				free(cookie);
+				if (cookie_with_prefix == NULL) {
+					log_error("Could not prepend \"SVPNCOOKIE=\" to the cookie.\n");
+					goto exit;
+				}
+
 				free(cli_cfg.cookie);
-				cli_cfg.cookie = cookie;
+				cli_cfg.cookie = cookie_with_prefix;
```
Comment (DimitriPapadopoulos) [inline src/main.c:550]: I would keep it simple as well, without checks, focusing on adding support for `SVPNCOOKIE=`:
```c
				char *cookie = read_from_stdin(COOKIE_SIZE);

				if (cookie == NULL) {
					log_warn("Could not read the cookie from stdin");
					break;
				}

				free(cli_cfg.cookie);
				cli_cfg.cookie = strdup_with_prefix(cookie, "SVPNCOOKIE=");
				free(cookie);
```

File: src/main.c [hunk 3]
```diff
@@ -517,19 +517,37 @@ int main(int argc, char **argv)
...
 			           "cookie-on-stdin") == 0) {
-				char *cookie = read_from_stdin(COOKIE_SIZE);
+				char *cookie;
+				char *cookie_with_prefix;
 
+				cookie = read_from_stdin(COOKIE_SIZE);
 				if (cookie == NULL) {
-					log_warn("Could not read the cookie from stdin");
-					break;
+					log_error("Could not read the cookie from stdin\n");
```
Comment (DimitriPapadopoulos) [inline src/main.c:538]: The extra `\n` seems wrong.

Why `log_warn` → `log_error`? I would keep that for a different PR if needed.

File: src/userinput.c [hunk 4]
```diff
@@ -363,3 +363,31 @@ char *read_from_stdin(size_t count)
...
+	buf = malloc(input_len + prefix_len);
+	if (buf == NULL)
+		return NULL;
+
+	memcpy(buf, prefix, prefix_len);
+	memcpy(buf + prefix_len, input, input_len);
+	buf[input_len + prefix_len] = '\0';
+
+	return buf;
+}
```
Comment (DimitriPapadopoulos) [inline src/userinput.c:393]: I would rather add this as a static function in `src/main.c`. Or otherwise `src/config`.

File: src/main.c (hunk 3 above)
Comment (gm-vm) [reply src/main.c]: `\n` is actually correct, the `log_*` functions do not add it.

Anyway, to me erroring out seems the best way to handle errors here, reason why I changed it from a warning to an error.

Agreed on keeping this out, given that I assume all of this is going to be squashed. I would have kept these changes separate, but a different PR probably makes sense.

File: src/userinput.c (hunk 4 above)
Comment (gm-vm) [reply src/userinput.c]: I turned it back into a static function, since I doubt it'll be needed elsewhere.

File: src/main.c [hunk 5]
```diff
@@ -517,19 +517,37 @@ int main(int argc, char **argv)
...
+					goto exit;
 				}
+
+				cookie_with_prefix = strdup_with_prefix(cookie, "SVPNCOOKIE=");
+				free(cookie);
```
Comment (DimitriPapadopoulos) [inline src/main.c]: I understand now why the initial function was a bit complex: to avoid double allocation if not needed (already starts with `SVPNCOOKIE=`). Perhaps you can check whether cookie starts with  `SVPNCOOKIE=` before re-allocating.

File: src/main.c [hunk 6]
```diff
@@ -529,7 +557,7 @@ int main(int argc, char **argv)
 					break;
 				}
 				free(cli_cfg.cookie);
-				cli_cfg.cookie = cookie;
+				cli_cfg.cookie = strdup_with_prefix(cookie, "SVPNCOOKIE=");
```
Comment (DimitriPapadopoulos) [inline src/main.c:560]: Just need to `free(cookie).

Perhaps better, with a test:
```c
				free(cli_cfg.cookie);
				if (strncmp(cookie, "SVPNCOOKIE=", strlen("SVPNCOOKIE=")) == 0) {
					cli_cfg.cookie = strdup_with_prefix(cookie, "SVPNCOOKIE=");
					free(cookie);
				} else  {
					cli_cfg.cookie = cookie;
				}
```
Not sure it's worth it though. Opinion?

File: src/main.c (hunk 3 above)
Comment (DimitriPapadopoulos) [reply src/main.c]: You're right of course, we can fix all error messages in a different PR.

File: src/main.c (hunk 1 above)
Comment (DimitriPapadopoulos) [reply src/main.c]: OK, I understand we can add `free(cookie)` in #1041, right? Perhaps in this case you can add it, because all the surrounding code is changed. Besides, for consistency, I think you should add `free(cookie);` and remove `free(cli_cfg.cookie);`.

Comment (gm-vm) [reply src/main.c]: Right, I removed too much stuff. I kept `free(cli_cfg.cookie);` simply because it was there already, but I can remove it to make the code consistent.

File: src/main.c (hunk 6 above)
Comment (gm-vm) [reply src/main.c:560]: Not having the checks make the code look a bit simpler and it avoids a tiny bit of duplication, given that it needs to be done for both `cookie` and `cookie-on-stdin`. And what about `strdup_with_prefix`? Should that blindly add the prefix and rely on callers?

In any case, I don't have strong opinions on this, so I'll let you decide. I don't think doing a tiny `malloc()` here for a short-lived string should really matter too much.

File: src/main.c (hunk 1 above)
Comment (DimitriPapadopoulos) [reply src/main.c]: Keep it if it was already there. Looks good.

File: src/main.c [hunk 7]
```diff
@@ -186,6 +186,27 @@ PPPD_USAGE \
...
+ * Returns the given "input" prefixed with "prefix" in a dynamically
+ * allocated string. This behaves exactly like "strdup" if "input" already
+ * starts with "prefix".
+ */
+static char *strdup_with_prefix(const char *input, const char *prefix)
+{
+	size_t prefix_len = strlen(prefix);
+	char *output;
+
+	if (memcmp(prefix, input, prefix_len) == 0)
```
Comment (gm-vm) [inline src/main.c:199]: Technically this allows `memcpy` to read `prefix_len` bytes starting from `input`, which is wrong in case `input` is shorter than that. This is why I was comparing the lengths of the strings. Maybe you wanted to use `strncmp`? I used the `mem*` functions simply because I needed to compute the length of everything already.

Comment (DimitriPapadopoulos) [reply src/main.c:199]: Right, `strcmp()` it is.
